File: bindings/python/tests/test_dolt_backend.py
import os
import unittest

from gnucash import Session, SessionOpenMode, ERR_BACKEND_NO_ERR
from gnucash import gnucash_core_c


class DoltBackendTestCase(unittest.TestCase):
    """Basic integration tests for the Dolt backend via Python bindings.

    These tests require TEST_DOLT_URL to be set to a valid Dolt-backed
    URL, e.g. "dolt://user:password@host/database-name/". If the
    variable is not set the tests are skipped.
    """

    @classmethod
    def setUpClass(cls):
        cls.url = os.environ.get("TEST_DOLT_URL")
        if not cls.url:
            raise unittest.SkipTest("TEST_DOLT_URL not set; skipping Dolt backend tests")

    def test_session_open_and_detection(self):
        with Session(self.url, SessionOpenMode.SESSION_NORMAL_OPEN) as sess:
            self.assertTrue(sess.is_dolt_backend())

    def test_branch_operations(self):
        with Session(self.url, SessionOpenMode.SESSION_NORMAL_OPEN) as sess:
            self.assertTrue(sess.is_dolt_backend())

            # Explicitly select the main branch so any Dolt operations
            # operate against a known branch instead of relying on the
            # server's default.
            sess.dolt_checkout_branch("main")

            # List branches; should not error, may be empty. We
            # intentionally avoid creating or checking out throwaway
            # branches so this test exercises only read-only branch
            # metadata.
            branches = sess.dolt_list_branches()
            self.assertIsInstance(branches, list)

    def test_commit(self):
        with Session(self.url, SessionOpenMode.SESSION_NORMAL_OPEN) as sess:
            self.assertTrue(sess.is_dolt_backend())

            # Explicitly select the main branch so add/commit operate
            # against a known branch instead of relying on the server's
            # default.
            sess.dolt_checkout_branch("main")

            # Ensure we can call add/commit without raising.
            sess.dolt_add()
            commit_hash = sess.dolt_commit("Test commit from Python bindings")
            # Commit hash may be None if not available; just assert the
            # call didn't raise.
            self.assertTrue(commit_hash is None or isinstance(commit_hash, str))

    def test_safe_save_flushes_without_error(self):
        # Open a Dolt-backed session without using the context manager so
        # we can control when safe_save() is invoked.
        sess = Session(self.url, SessionOpenMode.SESSION_NORMAL_OPEN)
        try:
            self.assertTrue(sess.is_dolt_backend())

            # Explicitly select the main branch so that safe_save()
            # operates against a known Dolt branch instead of relying
            # on the server's default.
            sess.dolt_checkout_branch("main")

            # Mark the book as having unsaved changes so that safe_save()
            # exercises the flush path (safe_save does not create Dolt
            # commits; callers must explicitly dolt_add/dolt_commit).
            book = sess.book
            gnucash_core_c.qof_book_mark_session_dirty(book.get_instance())

            # Invoke safe_save() directly; for Dolt this should flush any
            # pending changes to the working set on the explicitly
            # selected branch without leaving a backend error.
            sess.safe_save(None)
            self.assertEqual(sess.get_error(), ERR_BACKEND_NO_ERR)
        finally:
            sess.end()
            sess.destroy()


if __name__ == "__main__":
    unittest.main()

File: libgnucash/backend/dbi/test/test-dolt-backend.cpp
/********************************************************************\
 * test-dolt-backend.cpp: Basic tests for the Dolt backend          *
 *                                                                  *
 * These tests are intended to be run against an existing Dolt      *
 * database specified via the TEST_DOLT_URL environment variable,   *
 * e.g.:                                                            *
 *   TEST_DOLT_URL="dolt://user:password@host/database-name/"      *
 *                                                                  *
 * The tests will be skipped if TEST_DOLT_URL is not set.           *
 ********************************************************************/

#include <config.h>

#include <gtest/gtest.h>

#include "qof.h"
#include "qofsession.h"
#include "qofbook.h"
#include "qofbackend.h"

#include "gnc-backend-dbi.h"
#include "gnc-backend-dolt.h"

namespace {

class DoltBackendTest : public ::testing::Test
{
protected:
    void SetUp() override
    {
        /* Ensure the DBI backend and its libdbi instance are initialized
         * before attempting to open any Dolt-backed sessions. The GLib
         * test fixtures finalize the backend after each run, so we need
         * to re-initialize here for the GoogleTest-based suite. */
        gnc_module_init_backend_dbi();

        url = std::getenv("TEST_DOLT_URL");
        if (!url)
            GTEST_SKIP() << "TEST_DOLT_URL not set; skipping Dolt backend tests";
    }

    const char* url {nullptr};
};

TEST_F(DoltBackendTest, BasicSessionOpen)
{
    auto book = qof_book_new();
    auto session = qof_session_new(book);

    qof_session_begin(session, url, SESSION_NORMAL_OPEN);
    EXPECT_EQ(qof_session_get_error(session), ERR_BACKEND_NO_ERR);

    auto be = qof_session_get_backend(session);
    ASSERT_NE(be, nullptr);
    EXPECT_TRUE(gnc_dolt_backend_is_dolt(be));

    qof_session_end(session);
    qof_session_destroy(session);
}

TEST_F(DoltBackendTest, BranchOperations)
{
    auto book = qof_book_new();
    auto session = qof_session_new(book);

    qof_session_begin(session, url, SESSION_NORMAL_OPEN);
    EXPECT_EQ(qof_session_get_error(session), ERR_BACKEND_NO_ERR);

    auto be = qof_session_get_backend(session);
    ASSERT_NE(be, nullptr);
    ASSERT_TRUE(gnc_dolt_backend_is_dolt(be));

    // Explicitly select the main branch so that any subsequent Dolt
    // operations operate against a known branch instead of relying on
    // the server's default.
    ASSERT_TRUE(gnc_dolt_checkout_branch(be, "main"));

    // List branches; should not error, may be empty. We intentionally
    // avoid creating or checking out throwaway branches so that this
    // test exercises only read-only branch metadata on the main branch.
    auto branches = gnc_dolt_list_branches(be);
    if (branches)
    {
        for (gchar** p = branches; *p; ++p)
            g_free(*p);
        g_free(branches);
    }

    qof_session_end(session);
    qof_session_destroy(session);
}

TEST_F(DoltBackendTest, SafeSaveFlushesWithoutError)
{
    auto book = qof_book_new();
    auto session = qof_session_new(book);

    // Open a normal session on the default Dolt branch.
    qof_session_begin(session, url, SESSION_NORMAL_OPEN);
    ASSERT_EQ(qof_session_get_error(session), ERR_BACKEND_NO_ERR);

    auto be = qof_session_get_backend(session);
    ASSERT_NE(be, nullptr);
    ASSERT_TRUE(gnc_dolt_backend_is_dolt(be));

    // Explicitly select the main branch so that safe_save() operates
    // against a known Dolt branch instead of relying on the server's
    // default; GncDoltBackend::safe_sync now requires this.
    ASSERT_TRUE(gnc_dolt_checkout_branch(be, "main"));

    // Mark the book dirty to ensure safe_save() has something to flush.
    qof_book_mark_session_dirty(qof_session_get_book(session));

    // Call safe_save() directly; for Dolt this should safely flush any
    // pending changes into the working set on the explicitly selected
    // branch. Creating Dolt commits is an explicit caller operation and
    // is not performed by safe_save().
    qof_session_safe_save(session, nullptr);
    EXPECT_EQ(qof_session_get_error(session), ERR_BACKEND_NO_ERR);

    qof_session_end(session);
    qof_session_destroy(session);
}

} // anonymous namespace
